daemon/cluster/cluster.go

```diff
@@ -395,31 +395,32 @@ func (c *Cluster) Init(req types.InitRequest) (string, error) {
 
 	localAddr := listenHost
 
-	// If the advertise address is not one of the system's
-	// addresses, we also require a listen address.
-	listenAddrIP := net.ParseIP(listenHost)
-	if listenAddrIP != nil && listenAddrIP.IsUnspecified() {
+	// If the local address is undetermined, the advertise address
+	// will be used as local address, if it belongs to this system.
+	// If the advertise address is not local, then we need to detect
+	// a system address to use as listen and local addresses.
+	if net.ParseIP(localAddr).IsUnspecified() {
 		advertiseIP := net.ParseIP(advertiseHost)
-		if advertiseIP == nil {
-			// not an IP
-			c.Unlock()
-			return "", errMustSpecifyListenAddr
-		}
-
-		systemIPs := listSystemIPs()
 
 		found := false
-		for _, systemIP := range systemIPs {
+		for _, systemIP := range listSystemIPs() {
 			if systemIP.Equal(advertiseIP) {
+				localAddr = advertiseIP.String()
 				found = true
 				break
 			}
 		}
+
 		if !found {
-			c.Unlock()
-			return "", errMustSpecifyListenAddr
+			ip, err := c.resolveSystemAddr()
+			if err != nil {
+				c.Unlock()
+				logrus.Warnf("Could not find a local address: %v", err)
+				return "", errMustSpecifyListenAddr
+			}
+			listenHost = ip.String()
```

Comment:
Understand. But even today user is not allowed to pass `--listen-addr 0.0.0.0:<port>` and `--advertise-addr=<non local address>` combination.
The change takes effect only when adv addr is non local and we need a specific local address for the ipsec tunnels.

I probably don't see it, but if that seems the proper change to be done, I can take care of filling in only the local address. Let me know (only thing I will only be able to take care of it late tonight only, at the park now). Thanks.